ucp/_libs/arr.pyx

```diff
@@ -0,0 +1,287 @@
+from cpython.memoryview cimport (
+    PyMemoryView_FromObject,
+    PyMemoryView_GET_BUFFER,
+)
+from cpython.ref cimport Py_INCREF
+from cpython.tuple cimport PyTuple_New, PyTuple_SET_ITEM
+from cython cimport auto_pickle, boundscheck, wraparound
+from libc.stdint cimport uintptr_t
+from libc.string cimport memcpy
+
+try:
+    from numpy import dtype as numpy_dtype
+except ImportError:
+    numpy_dtype = None
+
+
+cdef dict itemsize_mapping = {
+    intern("|b1"): 1,
+    intern("|i1"): 1,
+    intern("|u1"): 1,
+    intern("<i2"): 2,
+    intern(">i2"): 2,
+    intern("<u2"): 2,
+    intern(">u2"): 2,
+    intern("<i4"): 4,
+    intern(">i4"): 4,
+    intern("<u4"): 4,
+    intern(">u4"): 4,
+    intern("<i8"): 8,
+    intern(">i8"): 8,
+    intern("<u8"): 8,
+    intern(">u8"): 8,
+    intern("<f2"): 2,
+    intern(">f2"): 2,
+    intern("<f4"): 4,
+    intern(">f4"): 4,
+    intern("<f8"): 8,
+    intern(">f8"): 8,
+    intern("<f16"): 16,
+    intern(">f16"): 16,
+    intern("<c8"): 8,
+    intern(">c8"): 8,
+    intern("<c16"): 16,
+    intern(">c16"): 16,
+    intern("<c32"): 32,
+    intern(">c32"): 32,
+}
+
+
+@auto_pickle(False)
+cdef class Array:
+    def __cinit__(self, obj):
+
+        cdef dict iface = getattr(obj, "__cuda_array_interface__", None)
+        self.cuda = (iface is not None)
+
+        cdef const Py_buffer* pybuf
+        cdef str typestr
+        cdef tuple data, shape, strides
+        cdef Py_ssize_t *shape_p
+        cdef Py_ssize_t *strides_p
+        cdef Py_ssize_t i
+        if self.cuda:
+            if iface.get("mask") is not None:
+                raise NotImplementedError("mask attribute not supported")
+
+            self.obj = obj
+
+            data = iface["data"]
+            self.ptr, self.readonly = data
+
+            typestr = iface["typestr"]
+            if typestr is None:
+                raise ValueError("Expected `str`, but got `None`")
+            elif typestr == "":
+                raise ValueError("Got unexpected empty `str`")
+            else:
+                try:
+                    self.itemsize = itemsize_mapping[typestr]
+                except KeyError:
+                    if numpy_dtype is not None:
+                        self.itemsize = numpy_dtype(typestr).itemsize
+                    else:
+                        raise ValueError(
+                            f"Unexpected data type, '{typestr}'."
+                            " Please install NumPy to handle this format."
+                        )
+
+            shape = iface["shape"]
+            strides = iface.get("strides")
+            self.ndim = len(shape)
+            if self.ndim > 0:
+                if strides is not None:
+                    if len(strides) != self.ndim:
+                        raise ValueError(
+                            "The length of shape and strides must be equal"
+                        )
+                    self.shape_p = <Py_ssize_t*>PyMem_Malloc(
```

Comment:
I know we are doing this to reduce overhead but maybe this is going a bit too far?
Could we use a regular python tuples for `self.shape` and `self.strides`?

Reply:
This just mirrors the same change that happened in PR ( https://github.com/rapidsai/ucx-py/pull/600 ).

FWIW this class still provides `shape` and `strides` as properties returning `tuple`s. So these remain accessible that way.

Reply:
Sorry I missed #600 but I still think that this is taking the low-level optimization too far. 
I mean ~500 ns should be negligible in this context or am I missing something?

Reply:
If we are ok with that overhead, what if went with something like Python's builtin `array`? That remains friendly to use from the Python side, but would also allow efficient typed access from the C-side.

Reply:
Great idea, best of both worlds :)

Reply:
Cool, have pushed changes along those lines.

Reply:
Just to note, this appears to add `~200ns` to `Array` creation time both for host and device cases. So this seems to be within what we consider negligible here.